projects/scannet_offline_eval/build_sparse_voxel_map.py

```diff
@@ -0,0 +1,475 @@
+                self.step(obs)
+            if self.cache_dir is not None:
+                # Save to cache
+                raise NotImplementedError
+        print(f"Found {len(self.voxel_map.get_instances())} instances")
+
+    ##############################################
+    # Language queries that return instances
+    ##############################################
+    def set_vocabulary(self, vocabulary: Dict[int, str]):
+        vocabulary = RearrangeDETICCategories(vocabulary)
+        self.semantic_sensor.update_vocabulary_list(vocabulary, SemanticVocab.SIMPLE)
+        self.semantic_sensor.set_vocabulary(SemanticVocab.SIMPLE)
+        return self.semantic_sensor.current_vocabulary
+
+    def get_instances_for_query(self, text_query: str) -> List[Instance]:
+        assert (
+            text_query in self.semantic_sensor.name_to_seg_id
+        ), f"{text_query} not in semantic_sensor vocabulary (current vocab: {self.semantic_sensor.current_vocabulary_id})"
+        query_class_id = self.semantic_sensor.name_to_seg_id[text_query]
+        instances = self.voxel_map.get_instances()
+        instances = [inst for inst in instances if inst.category_id == query_class_id]
+        return instances
+
+    def build_scene_and_get_instances_for_queries(
+        self, scene_obs: Dict[str, Any], queries: Sequence[str]
+    ):
+        """_summary_
+
+        Args:
+            scene_obs (Dict[str, Any]): Contains
+                - Images
+                - Depths
+                - Poses
+                - Intrinsics
+                - scan_name -- str that could be used for caching (but we probably also want to pass in dataset or sth in case we change resoluton, frame_skip, etc)
+            queries (Sequence[str]): _description_
+
+        Returns:
+            _type_: _description_
+        """
+        # Build scene representation
+        obs_list = []
+        for i in range(len(scene_obs["images"])):
+            obs = Observations(
+                gps=None,
+                compass=None,
+                rgb=scene_obs["images"][i] * 255,
+                depth=scene_obs["depths"][i],
+                semantic=None,
+                instance=None,  # These could be cached
+                # Pose of the camera in world coordinates
+                camera_pose=scene_obs["poses"][i],
+                camera_K=scene_obs["intrinsics"][i],
+                task_observations={
+                    # "features": scene_obs["images"][i],
+                },
+            )
+            obs_list.append(obs)
+        self.step_trajectory(obs_list)
+
+        # Do NMS on global bboxes
+        if self.global_nms_thresh > 0.0 and len(self.voxel_map.get_instances()) > 0:
+            # Loop this since when we combine bboxes, they are now bigger.
+            # And some bounding boxes might be inside the new larger bboxes
+            last_n_instances = -1
+            while len(self.voxel_map.get_instances()) != last_n_instances:
```

Comment:
Would this be useful to break out into a helper function in utils? Maybe that's too much

Reply:
Good idea -- I moved this to InstanceMemory to be clearer